Fix for Issue 31.

sroy4899 opened the pull request:
Fixed #31 (temporarily) by constraining all inline structs to be wild while allowing their fields to still be converted properly. 
Added 3 passing tests.

sroy4899 commented:
That was actually the first thing I tried before settling with the current solution, but I couldn't get it to work because I thought the only way to get access to the constraints was through `ProgramInfo`. But I combed over it this morning, and now I can get the fix working without any changes in `ProgramInfo.cpp`.  
I could also delete `LastRecordLoc` and have it be exactly what you specified if I were to use a handy function called `getPreviousDecl()` and then do some checks. The only issue is that `getPreviousDecl()` causes the tool to crash because it "is defined in multiple base classes" with differing types. This is doubly infuriating because `getNextDeclInContext(*Context)` works like a dream.
The most elegant solution I can think of now so that we don't always carry around this arbitrary `LastRecordLoc` is to create a local private boolean `InLineStructEncountered` in the `GlobalVisitor` class that will perform the check in `VisitRecordDecl`. I'll push up momentarily and you can take a look.

mwhicks1 commented:
Have you pushed yet, but I've missed it?

Once you have, I would move on to #84 so we can get that part done; this inline `struct` thing is lower priority.

sroy4899 commented:
The `code cleanup for fix` was that push! Sorry for the confusion. And yeah, I switched to a new branch and am already working on #84 :)

mwhicks1 commented on clang/lib/CConv/ConstraintBuilder.cpp:
```diff
@@ -311,13 +311,17 @@ class GlobalVisitor : public RecursiveASTVisitor<GlobalVisitor> {
+          Var->constrainToWild(Info.getConstraints(), Rsn);
+        }
+      }
     }
```
Don't you need to reset the `InlineStructEncountered` flag once you've processed definitions that use it?

sroy4899 replied:
I do! Thanks for pointing this out. I'll fix it now.

mwhicks1 replied:
For test cases: I would want to confirm that it works with
```
// a, b, c below all stay as WILD pointers
struct foo { int *x; } *a; // one decl; x rewrites to _Ptr<int>
struct bar { int* y; } *b, *c; // two decls, not one; y stays as int *
void foo(void) {
  a->x = (void*)0; 
  b->y = (void*)0;
  c->y = (int *)5; // forces it to WILD
}
```
You might also try a test where the `struct` is defined within a function, rather than globally.

sroy4899 replied:
Ok fixed!

mwhicks1 replied:
Are you sure this code is going to work with `struct bar { int* y; } *b, *c;`? Notice that there are two variables being declared here, not one. But it would seem that you are resetting the `InLineStructEncountered` eagerly, perhaps before the second variable is dealt with?

Also, I am not seeing a change to `VisitVarDecl`, i.e., for global variables. Did I miss it?

mwhicks1 replied:
Your test cases don't seem to have two variables being processed, at least not here.

You seem to also not have the case that there's a non-inline struct declared in between. This is what my test case was trying to show.

sroy4899 replied:
You're right in that the *c will actually get converted to a _Ptr. This might have been a misinterpretation on my part, but I think in a meeting you said the struct should only ever be defined once. So in the inline struct definition for b, struct bar is defined, and then when we come to c (which has now been split), there is no need to redefine struct bar, so c is allowed to be a _Ptr. It's clearly not consistent, so I understand the cause for concern. In fact, I'm just noting in your comment above that you expect C to stay WILD. Should I add support for C to become wild as well?

mwhicks1 replied:
Oh, I didn't realize that `c` was going to be split to its own definition. Interesting. Yes, for consistency (and to avoid confusion) I think it's better if they are all WILD, not just some of them.

sroy4899 replied:
Gotcha. I don't understand what you mean by no change to VisitVarDecl. I think I call the function processInLineStruct for global vars. Is it not there on your end?

sroy4899 replied:
That was another private static method I created because the constrain to wild occurs in both the function and global inline structs, so I abstracted it away into that function. Is that ok?

mwhicks1 replied:
I see. This is good -- thanks!

sroy4899 commented:
Sure, I forgot to refresh before commenting "fixed", so I didn't see your message. I can add them now

sroy4899 commented:
OK, thought I'd check in really quickly before pushing anything else. I can confirm the code snippet you posted above works. I can add it to the tests if you'd like, though I think my tests cover it. I had a concern about declaring structs inside functions though: 
When I run the tool on the following file: 
``` 
void foo(int *x) { 
    struct bar { int *x; } *y = 0;
}
``` 
I get the error `Expected a variable declaration but got an invalid AST node`. Weirdly, the file does compile.

mwhicks1 commented:
I guess we can just go with what you have, but maybe write up an issue for this.

sroy4899 commented:
Will do

sroy4899 commented:
Can we hold off on merging this? I want to check something really quick that I discovered when drafting up the issue.

sroy4899 commented:
Ok, the weird error I was getting where the tool complained about the AST is because RecordDecls aren't handled yet in the Function Visitor class in `ConstraintVariables.cpp`. When I went inside the FunctionVisitor and added support for RecordDecls, now everything works as expected and converts properly (by not converting the variable, but freely converting its fields). A lot of the code I added is the same code from the GlobalVisitor class. Should I add a method like `recordDeclSupport` or something to avoid the code duplication or should I hold off (since you said you want to minimize the number of extra stuff we add)?

sroy4899 commented:
New push. I scrapped the InLineStructEncountered and instead decided to carry out the LastRecordLoc, so no matter how many VarDecls are chained together after the RecordDecl, we can always check based on the last seen recorddecl and process iff the locations for both are the same

mwhicks1 commented on clang/lib/CConv/ConstraintBuilder.cpp:
```diff
@@ -17,6 +17,43 @@
 using namespace clang;
 
+private int lastRecordLocation = -1;
+private void processInlineStruct(VarDecl *VD, ProgramInfo &Info, ASTContext *Context) {
```
Seems to me you don't need this method anymore. In particular, you can just directly call `ConstraintResolver::constraintAllCVarsToWild` in the two places where you currently call this method, right?

sroy4899 commented:
Just made the change!
